File: src/lib/newsapi.ts
export interface NewsArticle {
  title: string;
  description: string;
  url: string;
  source: string;
  publishedAt: string;
  imageUrl: string | null;
}

export async function fetchAINews(): Promise<NewsArticle[]> {
  const apiKey = process.env.NEWSAPI_KEY;
  if (!apiKey) return [];

  const res = await fetch(
    `https://newsapi.org/v2/everything?q=artificial+intelligence+OR+LLM+OR+GPT+OR+Claude+OR+machine+learning&language=en&sortBy=publishedAt&pageSize=10`,
    {
      headers: { "X-Api-Key": apiKey },
      cache: "no-store",
    }
  );

  if (!res.ok) return [];

  const data = await res.json();
  return (data.articles || []).map((a: Record<string, unknown>) => ({
    title: a.title as string,
    description: a.description as string,
    url: a.url as string,
    source: (a.source as Record<string, string>)?.name || "Unknown",
    publishedAt: a.publishedAt as string,
    imageUrl: (a.urlToImage as string) || null,
  }));
}

File: server/app/clients/newsapi.py
import httpx

from app.config import settings
from app.schemas.news import NewsArticle


async def fetch_ai_news() -> list[NewsArticle]:
    if not settings.newsapi_key:
        return []

    url = (
        "https://newsapi.org/v2/everything"
        "?q=artificial+intelligence+OR+LLM+OR+GPT+OR+Claude+OR+machine+learning"
        "&language=en&sortBy=publishedAt&pageSize=10"
    )
    async with httpx.AsyncClient(timeout=10) as client:
        res = await client.get(url, headers={"X-Api-Key": settings.newsapi_key})
    if not res.is_success:
        return []

    articles = res.json().get("articles") or []
    return [
        NewsArticle(
            title=a.get("title", ""),
            description=a.get("description", ""),   
            url=a.get("url", ""),
            source=(a.get("source") or {}).get("name", "Unknown"),
            published_at=a.get("publishedAt", ""),
            image_url=a.get("urlToImage"),
        )
        for a in articles
    ]
